var/spack/repos/builtin/packages/rmgdft/package.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+from spack import *
+
+
+class Rmgdft(CMakePackage):
+    """RMGDFT is a high performance real-space density functional code
+       designed for large scale electronic structure calculations."""
+
+    homepage = "http://www.rmgdft.org/"
+    git      = "https://github.com/RMGDFT/rmgdft.git"
+    maintainers = ['elbriggs']
+    tags = ['ecp', 'ecp-apps']
+    version('master', branch='master')
+    version('4.2.1', tag='v4.2.1')
+
+    variant('build_type', default='Release',
+            description='The build type to build',
+            values=('Debug', 'Release', 'RelWithDebInfo'))
+
+    variant('qmcpack', default=True,
+            description='Build with qmcpack interface.')
+
+    variant('local_orbitals', default=True,
+            description='Build O(N) variant.')
+
+    # Normally we want this but some compilers (e.g. IBM) are
+    # very slow when this is on so provide the option to disable
+    variant('internal_pp', default=True,
+            description='Include built-in pseudopotentials. Normally '
+                        'enabled but some compilers are slow when '
+                        'this is on so provide a disable option.')
+
+    # RMGDFT 4.0.0 or later requires support for C++14
+    compiler_warning = 'RMGDFT 4.0.0 or later requires a ' \
+                       'compiler with support for C++14'
+    conflicts('%gcc@:4', when='@3.6.0:', msg=compiler_warning)
+    conflicts('%intel@:17', when='@3.6.0:', msg=compiler_warning)
+    conflicts('%pgi@:17', when='@3.6.0:', msg=compiler_warning)
+    conflicts('%llvm@:3.4', when='@3.6.0:', msg=compiler_warning)
+
+    depends_on('cmake@3.18.1', type='build')
+    depends_on('boost@1.74.0:')
+    depends_on('fftw-api@3')
+    depends_on('openblas threads=openmp')
```

Comment:
Do you really need the `openblas` package or just the BLAS API? Is OpenMP threading required?

Reply:
We require the BLAS API. Openblas is not specifically needed but we've had fewer build issues with it than with other blas libraries and the performance differences we've seen with MKL and ESSL have been small. OpenMP threading is not required but it can help with performance depending on the runtime parameters since RMG uses a hybrid model with MPI and threads. I debated not using it though since we have seen users set the number of OMP threads incorrectly via environment variables and over committing causes extremely poor performance.

Reply:
You don't want to overspecify in a spack package. The way to do this is to
`depends_on('blas')`

Then throw a conflict for each case that is known to be problematic, e.g. if you the ATLAS package was known to be a problem, you would do:
`conflict('atlas')`

You can see all the current providers, by type:
`spack providers blas`

Do you need LAPACK? if so, also add
`depends_on('lapack')`

Reply:
The choice of openblas was deliberate. As mentioned in an earlier comment RMG is a hybrid code. Getting the blas threading wrong can make it slow or even unusable. When built with openblas-openmp we control threads using the openblas api which normally works well (an exception being when the user overrides using env vars and makes a bad choice). I don't have access to all of the requisite platforms for testing the other 17 providers and given the range of build and thread control options available the potential for non performant outcomes is high.

We do use some LAPACK (and ScaLAPACK) but include a slimmed down version in the RMG source so there is no external dependence.

Reply:
2 reasons to leave the blas provider unspecified

1. Someone will create a faster blas. The spack package should express constraints, not preferences, because preferences will change
2. Someone will try to run this app with a tool that has an actual constraint on a different blas. They won't be able to, because they'll get a linker race. If you use `depends_on('blas')`, they can build with a different blas to make this work.

Could you try the following?
```
depends_on('blas')
depends_on('openblas threads=openmp', when='^openblas')
```

Reply:
> The choice of openblas was deliberate. As mentioned in an earlier comment RMG is a hybrid code. Getting the blas threading wrong can make it slow or even unusable. When built with openblas-openmp we control threads using the openblas api which normally works well (an exception being when the user overrides using env vars and makes a bad choice). I don't have access to all of the requisite platforms for testing the other 17 providers and given the range of build and thread control options available the potential for non performant outcomes is high.
> 
> We do use some LAPACK (and ScaLAPACK) but include a slimmed down version in the RMG source so there is no external dependence.

No need to test all 17 providers.

Reply:
> 2 reasons to leave the blas provider unspecified
> 
>     1. Someone will create a faster blas. The spack package should express constraints, not preferences, because preferences will change
> 
>     2. Someone will try to run this app with a tool that has an actual constraint on a different blas. They won't be able to, because they'll get a linker race. If you use `depends_on('blas')`, they can build with a different blas to make this work.
> 
> 
> Could you try the following?
> 
> ```
> depends_on('blas')
> depends_on('openblas threads=openmp', when='^openblas')
> ```

I like this suggestion.  

The only reason to do
`depends_on('openblas')`

Is if you know that the API is somehow incompatible with all other BLAS providers. For example, it only worked with CBLAS API and the ILP64 interface.

Reply:
I guess one thing I haven't made clear it's not just the API that matters, the threading model for the blas library is critical as well. RMG can run using a single thread/MPI or in a hybrid mode with multiple threads/MPI. Single thread/MPI works well for small problems/workstations and a serial BLAS library is fine. For large problems on supercomputers the hybrid mode with a threaded blas produces better performance as long as the thread count is set correctly. From reading the spack docs it looks like the system config can be set up to use a preferred provider for virtual packages like blas. The user has to know they need to do this though and how to make the correct choices. Even if a faster blas becomes available performance will likely be bad if the wrong threading model is selected. Openblas-openmp provides a way around this which may not always produce the best possible result but which usually gives a reasonable result for the full range of cases. I'm new to spack so perhaps I'm missing something but I don't see a way of using the virtual blas package to exclude packages based on their threading model so it appears it would need to be done for every one individually.

Reply:
This is my mindset: when putting together a spack package, I mostly don't pay attention to performance. I mostly focus on whether or not it can be built. I never typically exclude particularly combinations due to performance reason. It would take to long to test all these combinations. 

People who know what they are doing will pick the right blas as option. You can even put some comment in the Spack package about recommending a particular option.

Reply:
An HPC code not being concerned about performance seems a bit strange to me but leaving that aside it's not a question of testing every variant for performance since I know some of them perform badly. Is there any objection to me adding conflicts in cases where I know things will perform poorly? I ask since there will be users who don't know what they are doing.

Reply:
I think its fine to exclude certain combinations that perform poorly. I would strongly recommend using the conflict form with a `msg =  '...'` 

That include a meaningful message as to why this combination is excluded.